pkg/cli/mirror/catalog_images.go

```diff
@@ -233,6 +280,167 @@ func (o *MirrorOptions) processCatalogRefs(ctx context.Context, catalogsByImage
+	}
+	return opmCmdPath, nil
+}
+
+func extractOPMBinary(ctx context.Context, srcRef image.TypedImageReference, outDir string, insecure bool) error {
+	var img v1.Image
+	var err error
+	refExact := srcRef.Ref.Exact()
+	if srcRef.OCIFBCPath == "" {
+		remoteOpts := getCraneOpts(ctx, insecure)
+		img, err = crane.Pull(refExact, remoteOpts...)
+		if err != nil {
+			return fmt.Errorf("unable to pull image from %s: %v", refExact, err)
+		}
+	} else {
+
+		// obtain the path to where the OCI image reference resides
+		layoutPath := layout.Path(v1alpha2.TrimProtocol(srcRef.OCIFBCPath))
+
+		// get its index.json and obtain its manifest
+		rootIndex, err := layoutPath.ImageIndex()
+		if err != nil {
+			return err
+		}
+		rootIndexManifest, err := rootIndex.IndexManifest()
+		if err != nil {
+			return err
+		}
+
+		// attempt to find the first image reference in the layout...
+		// for a manifest list only search one level deep.
+
+	loop:
+		for _, descriptor := range rootIndexManifest.Manifests {
+
+			if descriptor.MediaType.IsIndex() {
+				// follow the descriptor using its digest to get the referenced index and its manifest
+				childIndex, err := rootIndex.ImageIndex(descriptor.Digest)
+				if err != nil {
+					return err
+				}
+				childIndexManifest, err := childIndex.IndexManifest()
+				if err != nil {
+					return err
+				}
+
+				// at this point, find the first image and store it for later if possible
+				//TODO extract the child index that corresponds to this machine's architecture
+				for _, childDescriptor := range childIndexManifest.Manifests {
+					if childDescriptor.MediaType.IsImage() && childDescriptor.Platform.Architecture == runtime.GOARCH && childDescriptor.Platform.OS == runtime.GOOS {
+						img, err = childIndex.Image(childDescriptor.Digest)
+						if err != nil {
+							return err
+						}
+						// no further processing necessary
+						break loop
+					}
+				}
+
+			} else if descriptor.MediaType.IsImage() {
+				// this is a direct reference to an image, so just store it for later
+				img, err = rootIndex.Image(descriptor.Digest)
+				if err != nil {
+					return err
+				}
+				// no further processing necessary
+				break loop
+			}
+		}
+	}
+	// if we get here and no image was found bail out
+	if img == nil {
+		return fmt.Errorf("unable to obtain image for %v", srcRef)
+	}
+	tr := tar.NewReader(mutate.Extract(img))
+	opmBinaryExtracted := ""
+	for {
+		header, err := tr.Next()
+
+		// break the infinite loop when EOF
+		if errors.Is(err, io.EOF) {
+			break
+		}
+
+		// skip the file if it is a directory or not in the bin dir
+		if !strings.HasSuffix(header.Name, opmBinarySuffix) || header.FileInfo().IsDir() {
+			continue
+		}
+
+		var buf bytes.Buffer
+		_, err = buf.ReadFrom(tr)
+		if err != nil {
+			return err
+		}
+
+		targetFileName := filepath.Join(outDir, header.Name)
+		bytes := buf.Bytes()
```

Comment:
Not sure where else to put this commentary, so including it here. There's a long line of different base images that have been used over time, and the opm executable could literally be anywhere. Here's a brief description of upstream and downstream images, to try to highlight some of the complexity you're going to run into.

## Upstream images

### Upstream images used over time (not an exhaustive list)

quay.io/operator-framework/upstream-opm-builder
quay.io/operator-framework/upstream-registry-builder 
quay.io/operator-framework/opm:latest                <- this is the current image


### Upstream possible paths (some of these could have been a symbolic link)

/opm      <- opm at root is one possibility!
/bin/opm

## Downstream image

### Downstream (not an exhaustive list)

registry.redhat.io/openshift4/ose-operator-registry:v4.13

### Downstream paths for 4.13

bin/darwin-amd64-opm -> /bin/registry/darwin-amd64-opm
bin/opm -> /bin/registry/opm
bin/windows-amd64-opm -> /bin/registry/windows-amd64-opm

---

Symbolic links are going to be the most difficult things to handle here. In theory you could have a opm symbolic link within the bin directory pointing to something outside the bin directory. None of the examples I show above do this, but it is something that COULD happen. Also the opm executable could be at the root.

At any rate, the section of code I have highlighted here should be checking if the number of bytes is zero. The reason for this is because according to the [tar read docs](https://pkg.go.dev/archive/tar#Reader.Read):

> Calling Read on special types like TypeLink, TypeSymlink, TypeChar, TypeBlock, TypeDir, and TypeFifo returns (0, io.EOF) regardless of what the Header.Size claims.

So your byte buffer could have zero bytes if the file is a symbolic link. In the downstream path examples above, you'd attempt to write out `bin/darwin-amd64-opm` with zero bytes and `/bin/registry/darwin-amd64-opm` which is a real executable which should have N number of bytes. So there's no point in writing out symbolic links since they won't really be useful.